### PR TITLE
Support Hooks [WIP]

### PR DESCRIPTION
**Pull request by christopherjbaker**
This is an attempt at #191. The old code is still in place, though it has been moved to the `legacy` folder.

### `useStore`

The primary use. Pass a `Store` constructor and props, returns and observable to use in the render.

```jsx
function TestComponent(props) {
  const { foo, bar } = useStore(Store, props);

  return <div>{foo}{bar}</div>;
}
```

### `useObserver`

An alternative use when you have an existing store and just need observability.

```jsx
function TestComponent() {
  useObserver();

  return <div>{store.foo}{store.bar}</div>;
}
```

### ModelProvider

Though the idea is not hook-specific, we have been discussing this for quite some time, so I decided to include it in the PR. Allows you to provide model classes at higher levels and consume them lower. Searches cascade up the tree.

> Obviously this is a contrived example, to show multiple definitions in a single provider, as well as nested providers.

```jsx
function TestComponent(props) {
  const Foo = useModel('Foo');
  const Bar = useModel('Bar');
  const Bam = useModel('Bam');

  const foo = useStore(Foo, props);
  const bar = useStore(Bar, props);
  const bam = useStore(Bam, props);

  return <div data-testid="foo">{foo.name} {bar.name} {bam.name}</div>;
}

function TestComponent(props) {
  return (
    <ModelProvider Foo={Foo}>
      <ModelProvider Bar={Bar} Bam={Bam}>
        <TestComponent />
      </ModelProvider>
    </ModelProvider>
  );
}
```

### CODE REVIEW

File: .eslintignore [hunk 1]
```diff
@@ -1,2 +1,3 @@
 dist
 node_modules
+legacy
```
Comment (BigAB) [inline .eslintignore:3]: Why do we ignore the `legacy` directory

File: .eslintrc.yml [hunk 2]
```diff
@@ -1,138 +1,66 @@
...
-  react/jsx-wrap-multilines:
+  spaced-comment:
     - error
+    - always
     -
-      declaration: true
-      assignment: true
-      return: true
-      arrow: true
+      exceptions: [ "!steal-remove-start", "!steal-remove-end" ]
```
Comment (BigAB) [inline .eslintrc.yml]: Can we just use the prettier rules for eslint and then we don't have to bike shed on code styles forever

File: lib/use-observer.js [hunk 3]
```diff
@@ -0,0 +1,21 @@
+import { useEffect, useState, useRef } from 'react';
+import Observer from './can-observer';
+
+export default function useObserver(name) {
+  const [ , update ] = useState();
```
Comment (BigAB) [inline lib/use-observer.js]: I wonder if we should also expose a "function decorator", similar to [`React.memo`](https://reactjs.org/blog/2018/10/23/react-v-16-6.html#reactmemo) where they can just make an observed component like:
```
import { observer } from 'ylem';
import Swanky3rdPartyComponent from 'swank-component-library'

export default observer(Swanky3rdPartyComponent)
```

which would just be a shorthand for 
```
export default props => {
   useObserver();
  return <Swanky3rdPartyComponent {...props} />
}
```

**Review by BigAB — APPROVED** [review]

**Comment by BigAB** [comment]
@christopherjbaker after reading https://overreacted.io/why-isnt-x-a-hook/ I am not sure our API choice is optimal.

Namely, the timing of when you call the `useStore()` matters, because it sets up listeners, so read before it aren't registered and opening a new observer closes the last one, which may break composition.

What are your thoughts?

**Comment by christopherjbaker** [comment]
TLDR: We are providing something that sits outside the norm of React, but I don't think we violate anything brought up in that article.

@BigAB That was a good read. I don't think we violate the issues brought up in that article though. `useStore` and `useObserver` do not break composition. I think it is perfectly reasonable that you cannot use observables until after you declare that you are going to, in the same way that you cannot use state until after you declare said state. You can absolutely `useStore` multiple times (this will create multiple observation readers under the hood, but that will not cause any issues), and debugging does not become any more complicated (other than what observables themselves add).

File: lib/use-observer.js (hunk 3 above)
Comment (christopherjbaker) [reply lib/use-observer.js]: We certainly could. Is this a pattern we want to encourage? If so, do we want to expose this *instead* of `useObserver`?

File: .eslintignore (hunk 1 above)
Comment (christopherjbaker) [reply .eslintignore:3]: this is just eslint ignores. Because the old code used a different rule set (that was far more complicated, btw. I switched to airbnb with just a few tweaks instead), and I didn't want to introduce a bunch of changes to that code in the git history.

File: .eslintrc.yml (hunk 2 above)
Comment (christopherjbaker) [reply .eslintrc.yml]: I will not use something that modified code for me, so we cannot use stock prettier. It does not look like the prettier rules are released in any other way besides that tool.

Before you ask, `standard` is not standard and I will not let go of semicolons.

As noted above, this project used to use a fully custom eslint config, and i switched it to use airbnb with some changes. I can revisit how many of those changes are just preferences and how many I think are necessary, in order to reduce the differences, but a few (like the one you specifically tagged here, because this is being built with steal) will have to stay.

File: lib/use-observer.js (hunk 3 above)
Comment (BigAB) [reply lib/use-observer.js]: I'm not sure... but yeah, I think we should, at least offer both. But I am open to being talked out of it.

File: .eslintrc.yml (hunk 2 above)
Comment (BigAB) [reply .eslintrc.yml]: > I will not use something that modified code for me

Why not? Isn't that kind of the point?

We could go with [eslint-plugin-prettier](https://github.com/prettier/eslint-plugin-prettier)
That way, anyone cool with it can let their editor auto fix, and those who don't like tools changing their code can just fix the issues themselves

**Review by mjstahl — APPROVED** [review]

**Review by mikedane — APPROVED** [review]

**Review by BigAB — APPROVED** [review]
